[Guo Kecheng] iP

Guo-KeCheng opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind from having to remember things you need to do. It's,

- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/Guo-KeCheng/ip/releases/tag/A-Jar).
2. double-click it.
3.  add your tasks.
4. let it manage your tasks for you 😉
5. And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
    public static void main(String[] args) {
        new Duke(FILE_PATH).run();
    }
}
```

Commit:
Add Gradle support

Guo-KeCheng committed:
Add Level 1-1 features

Guo-KeCheng committed:
Add Level 2-1 with comments

Guo-KeCheng committed:
Add Level 1-2 with comments

Guo-KeCheng committed:
Merge branch 'master' of https://github.com/Guo-KeCheng/ip

Guo-KeCheng committed:
Implement Level 3: Mark as Done

Guo-KeCheng committed:
Add level 4 features. However autotest needs to be updated

Guo-KeCheng committed:
Implement Level 5: Testing with documentations

Guo-KeCheng committed:
Implement Level 6: Delete

Guo-KeCheng committed:
Fix the Automated Testing

Guo-KeCheng committed:
Update input output txt file

Guo-KeCheng committed:
Update expected txt file

Guo-KeCheng committed:
Refactor commands into classes

Guo-KeCheng committed:
Implement save feature

Guo-KeCheng committed:
Implement Level-8: Datetime format

Guo-KeCheng committed:
Merge branch 'branch-Level-7'

Guo-KeCheng committed:
Merge and resolve merge conflict

Guo-KeCheng committed:
Nothng changed

Guo-KeCheng committed:
Merge branch 'add-gradle-support'

Guo-KeCheng committed:
Refactor to A-MoreOOP

Guo-KeCheng committed:
Implement packages

Guo-KeCheng committed:
Add JUnit tests

Guo-KeCheng committed:
Package the App as a JAR file

Guo-KeCheng committed:
Add Java Doc for task

Guo-KeCheng committed:
Add Java Doc for exception

Guo-KeCheng committed:
Tweak the code to comply with a coding standard

Guo-KeCheng committed:
Implement search feature

Guo-KeCheng committed:
Merge branch 'branch-A-CodingStandard'

Guo-KeCheng committed:
Resolve merge conflict when merging branch-Level-9

Guo-KeCheng committed:
Resolve merge conflict when merging branch-A-JavaDoc

Guo-KeCheng commented:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind from having to remember things you need to do. It's,

- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/Guo-KeCheng/ip/releases/tag/A-Jar).
2. double-click it.
3.  add your tasks.
4. let it manage your tasks for you 😉
5. And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
    public static void main(String[] args) {
        new Duke(FILE_PATH).run();
    }
}
```

Guo-KeCheng committed:
Use Checkstyle

Guo-KeCheng committed:
Merge branch 'branch-A-CheckStyle'

NappySprout commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,88 @@
+            String[] inputs = input.split(" ");
+            CommandType commandType = CommandType.valueOf(inputs[0].toUpperCase());
+
+            switch (commandType) {
```
very neat writing usage of switch statements 😄

NappySprout commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,91 @@
+ *
+ * @author Guo-KeCheng
+ */
+public abstract class Command {
```
Good abstraction of the classes under command 😁

NappySprout replied:
Good abstraction of the classes under command 😁

NappySprout commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,119 @@
+        Scanner sc = new Scanner(file); // create a Scanner using the File as the source
+
+        ui.printTaskExistence();
+        while (sc.hasNext()) {
```
Good job implementing saving and loading manually 😁

NappySprout commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,123 @@
+ *
+ * @author Guo-KeCheng
+ */
+public class Ui {
```
Very neat and tidy way of organising UI! but you will need to scrap this at level 10 😰

NappySprout replied:
Good way of laying out the UI, do you know how to incorporate this into level 10?

NappySprout reviewed (commented):
Looks really good, teh way you abstract out the files and split them out is very neat, the way you manually coded out the storage is amazing too.

NappySprout commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,119 @@
+     * @return TaskList with previously saved data
+     * @throws FileNotFoundException If source is not found
+     */
+    public TaskList load(String filePath) throws FileNotFoundException {
```
Good job implementing saving and loading manually 😁

NappySprout commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,88 @@
+     * @return command
+     * @throws DukeException Checks for invalid inputs
+     */
+    public static Command parse(String input, TaskList taskList, Ui ui, Storage storage)
```
Good use of switch statements and enums!

NappySprout reviewed (commented):
Good abstractions of the files and I am impressed with the way you managed your storage class, During level 10 some of the UI components can be reused so that might need some work when u reach there

Guo-KeCheng committed:
Implement basic GUI

Commit:
Add colour to GUI

Commit:
Merge branch 'branch-Level-10'

Guo-KeCheng committed:
Test

Guo-KeCheng committed:
Use var-args

Guo-KeCheng committed:
Use Assertions

There are no existing methods to check if assumptions are true
or events that are very unlikely to happen until there is a serious
flaw in the application.

Using assertions can help to keep track of any assumptions made.

Let's add assert to confirm validity of assumptions.

Guo-KeCheng committed:
Use Assertions

There are no existing methods to check if assumptions are true
or events that are very unlikely to happen until there is a serious
flaw in the application.

Using assertions can help to keep track of any assumptions made.

Guo-KeCheng committed:
Merge branch 'A-Assertions' of github.com:Guo-KeCheng/ip into A-Assertions

Guo-KeCheng committed:
A-CodeQuality

There are existing code that are not of the best code quality. An example
include the handling of invalid inputs. Error handling is done in individual
command classes causes issues when customising error messages.

Using the SLAP principle would allow the code to be more readable and
faciliate future changes.

Extracting out the error handling from individual command classes into
the parser class allows the error handling to be done in a centralised
location.

Guo-KeCheng committed:
Implement printList using streams

Guo-KeCheng committed:
Merge branch 'branch-A-CodeQuality'

Guo-KeCheng committed:
Resolve merge conflict

Guo-KeCheng committed:
Set up CI

Guo-KeCheng committed:
Resolve checkstyle errors

Guo-KeCheng committed:
Resolve checkstyle errors

Guo-KeCheng committed:
Merge branch 'master' of github.com:Guo-KeCheng/ip

Guo-KeCheng committed:
Merge branch 'master' of github.com:Guo-KeCheng/ip

Guo-KeCheng committed:
Merge branch 'master' of github.com:Guo-KeCheng/ip

Guo-KeCheng committed:
Add find free time extension

Guo-KeCheng committed:
Adjust tasklist

Guo-KeCheng committed:
Merge branch 'branch-B-FindFreeTimes'

Guo-KeCheng committed:
Improve the GUI

Guo-KeCheng committed:
Create _config.yml

Guo-KeCheng committed:
Update _config.yml

Guo-KeCheng committed:
Give a unique personality

Guo-KeCheng committed:
Merge branch 'master' of github.com:Guo-KeCheng/ip

Guo-KeCheng committed:
Create user guide

Guo-KeCheng committed:
Update README.md

Guo-KeCheng committed:
Update README.md

Guo-KeCheng committed:
Handle non number invalid input

Guo-KeCheng committed:
Resolve checkstyle issue and build jar file

Guo-KeCheng committed:
Update README.md

Added jar link

Guo-KeCheng committed:
Fix build errors and rebuild jar file

Guo-KeCheng committed:
Merge branch 'master' of github.com:Guo-KeCheng/ip

Guo-KeCheng committed:
Update README.md

Guo-KeCheng committed:
Add Ui.png

Guo-KeCheng committed:
Fix Documentation in Ui class
